### ISSUE TITLE
[Bug] Jailer fails to set cpuset

### ISSUE THREAD

**Issue by sameo** [firecracker-microvm/firecracker#2110]
## Describe the bug

When running a binary through jailer, I'm getting the following error:

```
./build/cargo_target/x86_64-unknown-linux-musl/release/jailer --id 5678 --exec-file /bin/pwd --gid 0 --uid 0 --node 0
Jailer error: Failed to write to /sys/fs/cgroup/cpuset/ls/5678/cpuset.cpus: Permission denied (os error 13)) = 107
```

This is *not* a generic jailer issue, but one specific to the machine I'm running this test on. See the next section section.

## Environment

`Firecracker v0.21.0-480-g0fe299df`
`Host system: x86-64`

This is a 2 NUMA node machine, 72 cores:

```
# lscpu 
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              72
On-line CPU(s) list: 0-71
Thread(s) per core:  2
Core(s) per socket:  18
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               63
Model name:          Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz
Stepping:            2
CPU MHz:             1197.852
CPU max MHz:         3600.0000
CPU min MHz:         1200.0000
BogoMIPS:            4588.98
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            46080K
NUMA node0 CPU(s):   0-17,36-53
NUMA node1 CPU(s):   18-35,54-71
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
```

Here come the interesting bits:

```
# cat /sys/fs/cgroup/cpuset/cpuset.cpus 
0,2-71

# cat /sys/devices/system/node/node0/cpulist 
0-17,36-53
```

Then this jailer piece of code: https://github.com/firecracker-microvm/firecracker/blob/master/src/jailer/src/cgroup.rs#L209-L216
will end up trying to write `0-17,36-53` to `/sys/fs/cgroup/cpuset/ls/5678/cpuset.cpus` which is not allowed because the parent cpuset does not include cpu 1 (It's made of `0,2-71`)

Some checks need to be applied before writing the NUMA cpu list back to the cpuset, e.g. by converting both the cpu list and the cpuset to cpu masks.

## Checks
- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [x] Are you certain the bug being reported is a Firecracker issue?

**Comment by sameo** [comment]
cc @alexandruag

**Comment by gc-plp** [comment]
Hi @sameo 

We plan on merging #2012 , which changes the behavior of the logic that you pointed out. Could you check if your usecase works with the changes provided in the PR?

**Comment by sameo** [comment]
> Hi @sameo
> 
> We plan on merging #2012 , which changes the the logic that you pointed out. Could you check if your usecase works with the changes provided in the PR?

iiuc https://github.com/firecracker-microvm/firecracker/pull/2012/files#diff-636ccd7e8fb64c33442e3c9d04519eadR131-R137 still does the same thing: Taking a NUMA node CPU list and directly applying it to the `cpuset.cpus` file. Is my understanding correct? If so, this won't fix the bug.

**Comment by christian7007** [comment]
Hello @sameo, the aim of https://github.com/firecracker-microvm/firecracker/pull/2012 is to achieve more flexibility on how the jailer manage the cgroups, so the users can accommodate more specific use cases. For your specific case you could do something like below, instead of using the `--node` attribute (which AFAIK will be deprecated at some point):

```
jailer ... --cgroup cpuset.mems=0 --cgroup cpuset.cpus=0,2-71
```

or probably something like this:

```
jailer ... --cgroup cpuset.mems=0 --cgroup cpuset.cpus=$(cat cat /sys/fs/cgroup/cpuset/cpuset.cpus)
```

**Comment by sameo** [comment]
> Hello @sameo, the aim of #2012 is to achieve more flexibility on how the jailer manage the cgroups, so the users can accommodate more specific use cases. For your specific case you could do something like below

Hey @christian7007. I am reporting that bug against jailer but I am seeing the issue when launching the Firecracker integration tests manually, where Firecracker is launched by jailer. I guess we'd have to fix those scripts to explicitly use the `--cgroup` argument.
I also believe that the command line:

```
--cgroup cpuset.mems=0 --cgroup cpuset.cpus=$(cat /sys/fs/cgroup/cpuset/cpuset.cpus)
```

should be the default behaviour when launching jailer without --cgroups arguments.

**Comment by christian7007** [comment]
> I guess we'd have to fix those scripts to explicitly use the --cgroup argument.

In #2012, the tests are modified for testing both, `--node` and `--cgroup`. For the `--node` tests, it just check `/sys/devices/system/node/nodeX/cpulist` so probably this will keep failing for your case. Even knowing that, I think tests are OK as the current behavior of `--node` is to set the content of  `/sys/devices/system/node/nodeX/cpulist` in the corresponding `cpuset.cpus`. So the functionality is correctly checked.

Probably it's up to the Firecracker maintainers team to decide if they want to elaborate the logic of `--node` to be able to support more complex scenarios or just allow the user to elaborate it himself by using `--cgroup` argument.

> should be the default behaviour when launching jailer without --cgroups arguments.

IMHO, this would reduce the flexibility (e.g in an environment where no isolation is required).

Also note that using `--cgroup cpuset.mems=X --cgroup cpuset.cpus=$(cat /sys/fs/cgroup/cpuset/cpuset.cpus)` could lead to an scenario where the process memory is isolated in node `X` while it can only use CPUs from other nodes, depending on the value of `/sys/fs/cgroup/cpuset/cpuset.cpus`.

**Comment by iulianbarbu** [comment]
@sameo , have you ran the integration tests on latest master? I've looked within the testing framework and we do not confine the jailer to a specific numa node or cpuset.cpus/mems [1, 2, 3]. At the same time the jailer does not confine the Firecracker process if not told so - `node` [4] and `cgroup` [5] args are optional, so you should be able to run the integration tests on your specific machine even if you run them confined to a subset of numa node resources.

Please let us know if the issue is still relevant to your use case.

[1] https://github.com/firecracker-microvm/firecracker/blob/master/tests/framework/jailer.py#L42
[2] https://github.com/firecracker-microvm/firecracker/blob/master/tests/framework/jailer.py#L48
[3] https://github.com/firecracker-microvm/firecracker/blob/master/tests/framework/microvm.py#L79
[4] https://github.com/firecracker-microvm/firecracker/blob/master/src/jailer/src/env.rs#L128
[5] https://github.com/firecracker-microvm/firecracker/blob/master/src/jailer/src/env.rs#L141

**Comment by alindima** [comment]
Taking a look over the comments, this is unrelated to the PR adding `--cgroups` parameter in Firecracker.

It is about pretty weird case where the cpus associated to the given numa node are not all included in the parent `cpuset` cgroup.
In the integration tests, we do run a test the `--node 0` configuration, which triggers this bug: https://github.com/firecracker-microvm/firecracker/blob/1f697a61e1fde68e1abb2e2ef75f7bf9acbc3440/tests/integration_tests/security/test_jail.py#L154

What we can do, as @sameo mentioned, is perform some extra checks to handle this case by restricting the cpuset to the minimal set of cpus that is at the intersection of cpulist of numa node X and the parent cgroup. I think this is the right path here, but I think it's of a pretty low priority, since it is a pretty odd corner case, and since it can be worked around by using the newer `--cgroup` parameter.

However, we've said for some time that we want to deprecate the `--node` parameter, since we've introduced `--cgroup`. I've opened the following issue to track this: https://github.com/firecracker-microvm/firecracker/issues/2613. I propose we start the deprecation, by adding a runtime warning when using it, with indications for using the new parameter, and saying that it will be deprecated in a future release.
We won't state which release just yet, since we don't have a deprecation policy in place yet, but we're working on it.

WDYT?
